concordium-consensus/src/Concordium/ImportExport.hs

```diff
@@ -736,19 +833,25 @@ writeChunks
             runPutH (liftPut $ putWord64be sectionHeaderLength >> put placeholderSectionHeader) chunkHdl
             blocksStart <- hTell chunkHdl
             return (sectionStart, blocksStart)
-        (sectionBlockCount, lastFinalizationRecordIndex') <- exportBlocksToChunk chunkHdl sectionFirstBlockHeight chunkSize lastFinalizationRecordIndex getBlockAt
+        (sectionBlockCount, lastFinalizationRecordIndex', lastExportedBlock) <- exportBlocksToChunk chunkHdl sectionFirstBlockHeight chunkSize lastFinalizationRecordIndex getBlockAt
         blocksEnd <- liftIO $ hTell chunkHdl
         -- Only write finalization records to a chunk if it's the last one for the section
         let lastExportedBlockHeight = sectionFirstBlockHeight + BlockHeight sectionBlockCount - 1
-        sectionFinalizationCount <-
+        (sectionFinalizationCount, sectionFinalizationEntryPresent) <-
             if lastExportedBlockHeight < sectionLastBlockHeight
-                then return 0
-                else exportFinRecsToChunk chunkHdl lastFinalizationRecordIndex' getFinalizationRecordAt
+                then return (0, False)
+                else do
+                    finRecs <- exportFinRecsToChunk chunkHdl lastFinalizationRecordIndex' getFinalizationRecordAt
+                    finEntryPresent <- exportFinalizationEntryToChunk chunkHdl lastExportedBlock getFinalizationEntry
+                    return (finRecs, finEntryPresent)
         liftIO $ do
             sectionEnd <- hTell chunkHdl
             -- Go back to the start and rewrite the section header with the correct data
             hSeek chunkHdl AbsoluteSeek sectionStart
-            let sectionHeader =
+            let sectionVersion = case sing @cpv of
+                    SConsensusParametersVersion0 -> SHV0
+                    SConsensusParametersVersion1 -> SHV1
```

Comment:
Isn't this going to export sections with version 1 for protocol 6?

So we will run this exporter now and it'll produce sections V1. Then node version 6.0.4 will not be able to import them even if it can fully understand them and only the very last section actually needs the finalization entry.

Wouldn't it make more sense to indicate whether the section value is actually different?

Reply:
Thinking on it again, I think it would be better to reuse the `sectionFinalizationCount` to and protocol version together and in the import do
```
                        if sectionProtocolVersion <= P5 then
                            replicateM_
                                (fromIntegral sectionFinalizationCount)
                                ( importData hdl fileSize $
                                    ImportFinalizationRecord sectionProtocolVersion sectionGenesisIndex
                                )
                        else when (sectionFinalizationCount == 0) $ do
                            importData hdl fileSize $
                                ImportFinalizationEntry sectionProtocolVersion sectionGenesisIndex
```

and then when exporting set the finalziation count section to 0 if there are any exports and 1 otherwise. This will make the exports produced by version 6.1 node compatible with those of 6.0 as well unless I am mistaken.

As long as we document the format we should be fine maintainability wise as well.

You could perhaps rename the variables, change the type to make it reflect the new setup.